server/cache.go

```diff
@@ -91,394 +89,144 @@ func (s *storesInfo) getStoreCount() int {
-}
-
-// getRegionByID gets the region and leader peer by regionID.
-func (r *regionsInfo) getRegionByID(regionID uint64) (*metapb.Region, *metapb.Peer) {
-	r.RLock()
-	defer r.RUnlock()
-
+func (r *regionsInfo) getRegion(regionID uint64) *regionInfo {
 	region, ok := r.regions[regionID]
 	if !ok {
-		return nil, nil
-	}
-
-	leaderStoreID, ok := r.leaders.regionStores[regionID]
-	if ok {
-		return cloneRegion(region), leaderPeer(region, leaderStoreID)
+		return nil
 	}
-
-	return cloneRegion(region), nil
+	return region.clone()
 }
 
-// getRegions gets all the regions, returns nil if not found.
-func (r *regionsInfo) getRegions() []*metapb.Region {
-	r.RLock()
-	defer r.RUnlock()
-
-	regions := make([]*metapb.Region, 0, len(r.regions))
-	for _, region := range r.regions {
-		regions = append(regions, cloneRegion(region))
+func (r *regionsInfo) setRegion(region *regionInfo) {
+	if origin, ok := r.regions[region.GetId()]; ok {
+		r.removeRegion(origin)
 	}
-
-	return regions
+	r.addRegion(region)
 }
 
-func (r *regionsInfo) addRegion(region *metapb.Region) {
-	r.searchRegions.insert(region)
-
-	_, ok := r.regions[region.GetId()]
-	if ok {
-		log.Fatalf("addRegion for already existed region in regions - %v", region)
-	}
-
+func (r *regionsInfo) addRegion(region *regionInfo) {
+	// Add to tree and regions.
+	r.tree.update(region.Region)
 	r.regions[region.GetId()] = region
 
-	r.addRegionCount(region)
-}
-
-func (r *regionsInfo) updateRegion(region *metapb.Region) {
-	r.searchRegions.update(region)
-
-	oldRegion, ok := r.regions[region.GetId()]
-	if !ok {
-		log.Fatalf("updateRegion for none existed region in regions - %v", region)
+	if region.Leader == nil {
+		return
 	}
 
-	r.regions[region.GetId()] = region
-
-	r.addRegionCount(region)
-	r.removeRegionCount(oldRegion)
-}
-
-func (r *regionsInfo) removeRegion(region *metapb.Region) {
-	r.searchRegions.remove(region)
-
-	_, ok := r.regions[region.GetId()]
-	if !ok {
-		log.Fatalf("removeRegion for none existed region in regions - %v", region)
+	// Add to leaders and followers.
+	for _, peer := range region.GetPeers() {
+		storeID := peer.GetStoreId()
+		if peer.GetId() == region.Leader.GetId() {
+			// Add leader peer to leaders.
+			store, ok := r.leaders[storeID]
+			if !ok {
+				store = make(map[uint64]*regionInfo)
+				r.leaders[storeID] = store
+			}
+			store[region.GetId()] = region
+		} else {
+			// Add follower peer to followers.
+			store, ok := r.followers[storeID]
+			if !ok {
+				store = make(map[uint64]*regionInfo)
+				r.followers[storeID] = store
+			}
+			store[region.GetId()] = region
+		}
 	}
+}
 
+func (r *regionsInfo) removeRegion(region *regionInfo) {
+	// Remove from tree and regions.
+	r.tree.remove(region.Region)
 	delete(r.regions, region.GetId())
 
-	r.leaders.remove(region.GetId())
-
-	r.removeRegionCount(region)
-}
-
-func (r *regionsInfo) addRegionCount(region *metapb.Region) {
+	// Remove from leaders and followers.
 	for _, peer := range region.GetPeers() {
```

Comment:
Can we meet following case:

1, add region a with peers 1,2,3,4
2, remove region a with peers 1,2,3

So the followers cache may have a uncleared peer.